### PR TITLE
feat: Add set states

### PR DESCRIPTION
**Pull request by dehilsterlexis**

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith the [update](https://github.com/hpcc-systems/Visualization/blob/6e6fc8c6ad7296c2e964909e6119de8758b01f8a/packages/comms/src/ecl/workunit.ts#L252) function uses the [Request](https://github.com/hpcc-systems/Visualization/blob/6e6fc8c6ad7296c2e964909e6119de8758b01f8a/packages/comms/src/services/wsWorkunits.ts#L1142) interface that does not have a StateID numeric argument, only a State string.

**Comment by GordonSmith** [comment]
@dehilsterlexis ECL Watch calls WUAction with:
```
Wuids: W20240118-144332
WUActionType: SetToFailed
ActionType: SetToFailed
```

**Comment by dehilsterlexis** [comment]
@GordonSmith SetToFailed is the only state changing action right now. There are 12 of them in total. Either we implement each one in the same way or look into creating a more generic SetState function. My suggestion is to add a new action called "SetState" in comms and in [onDFUWorkunitsAction](https://github.com/hpcc-systems/HPCC-Platform/blob/3bca1feae8ee06416001b6373095752b4ca92f5e/esp/services/ws_fs/ws_fsService.cpp#L1706). That is something I believe I could do in all three places (the third being vscode-ecl).

**Comment by GordonSmith** [comment]
The generic method is "WUAction" which takes an array of wuids and the ActionType.

**Comment by dehilsterlexis** [comment]
> The generic method is "WUAction" which takes an array of wuids and the ActionType.

Yes, but the way it is currently set up, we can only do set failed.

**Comment by GordonSmith** [comment]
I think they are all already implemented here:  https://github.com/hpcc-systems/Visualization/blob/trunk/packages/comms/src/ecl/workunit.ts#L365-L407 ?

**Comment by dehilsterlexis** [comment]
@GordonSmith those are not them. Here is the list: [click here](https://github.com/hpcc-systems/eclide/blob/704ba27413cb775ac2fa839a50ebb2b038b4ac92/eclide/WorkunitFrame.h#L181)

File: src/ecl/eclWatchTree.ts [hunk 1]
```diff
@@ -93,6 +93,21 @@ export class ECLWatchTree extends Tree {
...
+        });
+
+        vscode.commands.registerCommand("hpccPlatform.unprotectWU", (wuNode: ECLWUNode) => {
+            wuNode.unprotect();
+            this.refresh();
+        });
+
+        vscode.commands.registerCommand("hpccPlatform.setStateFailed", (wuNode: ECLWUNode) => {
+            wuNode.setStateFailed();
+            this.refresh();
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:108]: Remove `this.refresh`

File: src/ecl/eclWatchTree.ts [hunk 2]
```diff
@@ -400,6 +415,18 @@ export class ECLWUNode extends Item<ECLWatchTree> {
         this._wu.abort().then(() => this._tree._onDidChangeTreeData.fire(this));
     }
 
+    protect() {
+        this._wu.protect();
+    }
+
+    unprotect() {
+        this._wu.unprotect();
+    }
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:424]: Protect / unprotect from diff PR

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: src/ecl/eclWatchTree.ts [hunk 3]
```diff
@@ -400,6 +415,18 @@ export class ECLWUNode extends Item<ECLWatchTree> {
...
+
+    setStateFailed() {
+        this._wu.update({"State": "failed"});
+    }
+
```
Comment (GordonSmith) [inline src/ecl/eclWatchTree.ts:429]: Try this:
```

    setState(stateID: WUStateID) {
        const service = new WorkunitsService({ baseUrl: this._wu.BaseUrl });
        return service.WUUpdate({
            Wuid: this._wu.Wuid,
            State: stateID as unknown as string
        });
    }

    setStateFailed() {
        this.setState(WUStateID.Failed).then(() => this._tree.refresh());

```

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: package.json [hunk 4]
```diff
@@ -652,7 +712,71 @@
         }
       }
     ],
+    "submenus": [
+      {
+          "id": "sequencetype",
+          "label": "New Pass"
+      },
```
Comment (GordonSmith) [inline package.json:719]: What is this for?

File: package.json [hunk 5]
```diff
@@ -652,7 +712,71 @@
...
+      {
+          "id": "librarypass",
+          "label": "Library Pass"
+      },
```
Comment (GordonSmith) [inline package.json:723]: What is this for?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (dehilsterlexis) [reply package.json:719]: Copied it from vscode-nlp and didn't delete it.

File: package.json (hunk 5 above)
Comment (dehilsterlexis) [reply package.json:723]: Copied it from vscode-nlp and didn't delete it.

File: src/ecl/eclWatchTree.ts [hunk 6]
```diff
@@ -400,6 +448,62 @@ export class ECLWUNode extends Item<ECLWatchTree> {
...
+    setState(stateID: WUStateID) {
+        const service = new WorkunitsService({ baseUrl: this._wu.BaseUrl });
+        return service.WUUpdate({
+            Wuid: this._wu.Wuid,
+            State: stateID as unknown as string
+        });
+    }
+
+    setStateCompiled() {
+        this.setState(WUStateID.Compiled).then(() => this._tree.refresh());
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply src/ecl/eclWatchTree.ts:460]: Agreed

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**